Add virtual scrolling to SelectControl and country dropdowns

ilyasfoo opened the pull request:
### Changes proposed in this Pull Request:

Addresses p1740728084833759-slack-C01SFMVEYAK:

> When i skip onboarding and it goes to the country selector, I go to interact with the dropdown, type a country or open it and in those initial couple of seconds it performs doggishly slow and is painful to interact with. (Chromium browser)

This PR:

- Adds `react-window` to `@woocommerce/packages`
- Adds virtual scrolling support for `SelectControl` component
- Implements virtual scrolling for core profiler business step, skip profiler, and StoreLocation component

### Screenshots or screen recordings:

| Before | After |
| ------ | ----- |
|   <img src="https://github.com/user-attachments/assets/0e5360d3-fb0b-415a-9c35-7a54bbf6df16" width="300">|   <img src="https://github.com/user-attachments/assets/3a00fd64-eb40-4f5c-8461-ef547549aab6" width="300">   |


<!-- Uploading d.pr/i/j8rnyX to GitHub… -->

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Testing core profiler business step

1. In a fresh site, go to core profiler `/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard`
2. Click on `Set up my store` and `Continue`
3. Click on `Where is your store located` dropdown and observe the time it takes to show options is significantly faster
4. Try scrolling, selecting, searching, and other interactions and ensure it's working similarly to previous version (the extremely brief blank options while scrolling is expected due to virtual rendering) 

#### Testing core profiler skip profiler

1. Go to skip guided setup, `/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard&step=skip-guided-setup`
2. Click on `Select country/region` dropdown and observe similar experience as previous test

#### Testing StoreLocation component

1. Go to `WooCommerce > Home`
2. Click on `Collect sales tax`
3. Select Woo tax partner if asked
4. Click on `Country / Region` and observe similar experience as previous test

#### Testing storybook

1. Go to `tools/storybook`
2. Run `npm run watch:build:storybook`
3. Open storybook and navigate to `Components > SelectControl`
4. Test `Virtual scrolling with many options` and observe similar experience as previous test
5. Optionally, test other dropdowns and ensure they're working as expected

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ilyasfoo committed:
Add react-window, added SelectControl support for virtual list

ilyasfoo committed:
Add lockfile

ilyasfoo committed:
Fix support for initial select

ilyasfoo committed:
Set virtual SelectControl for skip wizard

ilyasfoo committed:
Changelogs

ilyasfoo committed:
Add virtual scrolling in StoreLocation component

ilyasfoo committed:
Add storybook

ilyasfoo committed:
Update readme

ilyasfoo committed:
Merge remote-tracking branch 'origin/trunk' into add/onboarding-business-step-virtual-select-control

ilyasfoo committed:
Remove comment

ilyasfoo committed:
Simplify props

chihsuan commented on packages/js/components/src/select-control/index.tsx:
```diff
@@ -229,6 +244,13 @@ export class SelectControl extends Component< Props, State > {
+		if ( selected !== prevProps.selected ) {
+			this.reset( selected );
+		}
+	}
```
Is it intended to do this for all cases? Or should we only reset the state when `virtualScroll` is `true`?

ilyasfoo replied:
Yes this is intended for all cases. Sorry I forgot to mention, this is a fix for initial selection where the pre-selected option is not highlighted when after a fresh reload.

chihsuan reviewed (commented):
Nice work!  @ilyasfoo  I found an issue with keyboard navigation. The selected option is not focused in the dropdown when using keyboard to navigate. 

See videos below.


Before:

https://github.com/user-attachments/assets/c3f9484b-4f61-40c8-86b0-e2ca5045cd34

After:

https://github.com/user-attachments/assets/a971311b-a1b4-4bbc-8a2b-063ed19145bc


Summarizing comment about my review goes here.

#### Review checklist

| Parameter  | Result |
| ------------- | ------------- |
| Approach  | ✅ All good |
| Implementation  | ✅ All good |
| Automated tests  | If it's possible, we can add some unit tests for the virtual scrolling |
| Naming  | ✅ All good |
| Typing  | ✅ |
| Code comments  | Left a comment in the PR |
| Changelog  | ✅ |
| Documentation  | ✅ |
| PR description  | ✅ All good |
| Testing instructions  | ✅ All good |
| Manual testing  | An issue found with keyboard navigation |

#### Scenarios tested

On a fresh JN site.


https://github.com/user-attachments/assets/5e807d13-a5db-4db0-b341-b99d912accf8

| Scenario  | Result | Screenshot |
| ------------- | ------------- | ------------- |
| Testing core profiler business step  | ✅ Functionality worked correctly and confirm it's faster with virtual scrolling |  |
| Testing core profiler skip profiler  |  ✅ Functionality worked correctly and confirm it's faster with virtual scrolling |  |
| Testing StoreLocation component  | ✅ Functionality worked correctly and confirm it's faster with virtual scrolling  |
| Storybook  | ✅|  |

chihsuan committed:
Merge branch 'trunk' into add/onboarding-business-step-virtual-select-control

ilyasfoo committed:
Fix tests

ilyasfoo committed:
Merge remote-tracking branch 'origin/trunk' into add/onboarding-business-step-virtual-select-control

ilyasfoo committed:
Fix last item hidden under viewport

ilyasfoo committed:
Fix test

ilyasfoo committed:
Add unit test for virtual scrolling

ilyasfoo committed:
Revert for better stability on non-clean local testing

ilyasfoo committed:
Merge remote-tracking branch 'origin/trunk' into add/onboarding-business-step-virtual-select-control

ilyasfoo committed:
Remove redundant cases

ilyasfoo commented:
Thanks for the feedback, @chihsuan!

> The selected option is not focused in the dropdown when using keyboard to navigate.

Good catch, I've fixed this in 5d9035917e732ed47c352558be82fbdaca6491e9. Tbh I have don't know exactly why though, I suspect it could be the list height is somehow larger than the viewport.

> If it's possible, we can add some unit tests for the virtual scrolling

Good call, I've added unit tests in 4c7928cbb66d616ed355d68ea5e44ffd2d964b15

moon0326 reviewed (approved):
LGTM and tested well 👍 🚀 

The list appears instantly now :)

ilyasfoo committed:
Merge branch 'trunk' into add/onboarding-business-step-virtual-select-control

chihsuan reviewed (approved):
Thanks @ilyasfoo, I verified that the keyboard navigation is functioning correctly. 👍 

I also checked the VoiceOver accessibility feature and noticed that because we only render a portion of the list, VoiceOver does not announce the total number of options but only the currently visible ones. I'm not sure if this is something we can address, but I wanted to bring it to your attention.

<img width="1147" alt="Screenshot 2025-03-04 at 10 28 27" src="https://github.com/user-attachments/assets/3fa9127f-c0f6-4d7c-a8f0-9fce7c1c4c56" />

ilyasfoo committed:
Merge remote-tracking branch 'origin/trunk' into add/onboarding-business-step-virtual-select-control

ilyasfoo committed:
Add aria attributes for virtual option to support accessibility

ilyasfoo commented:
Great catch, @chihsuan! Thanks for being so thorough. I added relevant aria attributes to support VoiceOver in d040156
